src/main/java/org/apache/sling/scripting/sightly/impl/utils/ScriptDependencyResolver.java

```diff
@@ -74,60 +76,100 @@ public class ScriptDependencyResolver implements ResourceChangeListener, Externa
+            cacheEnabled = true;
             resolutionCache = new Cache(cacheSize);
         }
-        componentContext.getBundleContext().addBundleListener(this);
+        if (cacheEnabled) {
+            componentContext.getBundleContext().addBundleListener(this);
+            Dictionary<String, Object> resourceChangeListenerProperties = new Hashtable<>();
+            resourceChangeListenerProperties.put(ResourceChangeListener.PATHS, ".");
+            resourceChangeListenerProperties.put(ResourceChangeListener.CHANGES, new String[] {
+                    ResourceChangeListener.CHANGE_ADDED,
+                    ResourceChangeListener.CHANGE_CHANGED,
+                    ResourceChangeListener.CHANGE_REMOVED
+            });
+            resourceChangeListenerServiceRegistration =
+                    componentContext.getBundleContext().registerService(ResourceChangeListener.class, this, resourceChangeListenerProperties);
+        }
+    }
+
+    @Deactivate
+    private void deactivate(ComponentContext componentContext) {
+        if (resourceChangeListenerServiceRegistration != null) {
+            resourceChangeListenerServiceRegistration.unregister();
+        }
+        componentContext.getBundleContext().removeBundleListener(this);
     }
 
     public Resource resolveScript(RenderContext renderContext, String scriptIdentifier) {
+        SlingHttpServletRequest request = BindingsUtils.getRequest(renderContext.getBindings());
+        if (!cacheEnabled) {
+            return internalResolveScript(request, renderContext, scriptIdentifier);
+        }
         readLock.lock();
         try {
-            SlingHttpServletRequest request = BindingsUtils.getRequest(renderContext.getBindings());
             String cacheKey = request.getResource().getResourceType() + ":" + scriptIdentifier;
             Resource result = null;
             if (!resolutionCache.containsKey(cacheKey)) {
                 readLock.unlock();
                 writeLock.lock();
                 try {
-                    Resource caller =
-                            ResourceResolution.getResourceForRequest(scriptingResourceResolverProvider.getRequestScopedResourceResolver(),
-                                    request);
-                    result = ResourceResolution.getResourceFromSearchPath(caller, scriptIdentifier);
-                    if (result == null) {
-                        SlingScriptHelper sling = BindingsUtils.getHelper(renderContext.getBindings());
-                        if (sling != null) {
-                            caller = getResource(scriptingResourceResolverProvider.getRequestScopedResourceResolver(),
-                                    sling.getScript().getScriptResource());
-                            result = ResourceResolution.getResourceFromSearchPath(caller, scriptIdentifier);
-                        }
-                    }
+                    result = internalResolveScript(request, renderContext, scriptIdentifier);
```

Comment:
this call is done with the writeLock held, so it's serialized. Would it be possible to run this code outside of the writeLock and do only the update of the resolutionCache with the writeLock held?

Reply:
How would that look like? You could push a commit to this branch, but I don't understand how you could write the code differently.

Reply:
I am thinking of something like this:

```
if (!resolutionCache.containsKey(cacheKey)) {
  readLock.unlock();
  result = internalResolveScript(request, renderContext, scriptIdentifier);
  writeLock.lock();
  try {
    if (result != null) {
      resolutionCache.put(cacheKey, result.getPath());
    } else {
      resolutionCache.put(cacheKey, NOT_FOUND_MARKER);
    }
    readLock.lock();
  } finally {
    writeLock.unlock();
  }
}
```

Reply:
I think this could lead to more threads reading the same resource in the case of concurrent requests for the same script dependency resolution.